### DISCUSSION TITLE
Adding back ["tagger", "parser", "ner", "attribute_ruler", "lemmatizer"] to a model

### DISCUSSION THREAD

**Discussion by Phat-Loc** [explosion/spaCy#7967] [Help: Coding & Implementations] [answered]
What I want to achieve is a model with a custom trained senter. I have created the training data for senter and am able to train it. However I now want to add back  ["tagger", "parser", "ner", "attribute_ruler", "lemmatizer"]  to the model. How do I achieve that? The following was the steps I took

1. nlp = spacy.load('en_core_web_md', exclude=["tagger", "parser", "ner", "attribute_ruler", "lemmatizer"]) this gives a model with just tok2vec pipeline
2. I created a doc_bin with training data from the nlp model
3. In my config.cfg I set the pipeline ["tok2vec","senter"]

Now that I trained the model with a custom senter how can I add back the rest of the pipeline? Or am I going about it the wrong way i.e. I should keep the other parts of the pipeline but just disable it for training? In other words  keep default pipeline i.e. disable instead of exclude and set in config.cfg `frozen_components = ["tagger", "parser", "ner", "attribute_ruler", "lemmatizer"]`? Do I also set the pipe `pipeline = ["tok2vec","senter", "tagger", "parser", "ner", "attribute_ruler", "lemmatizer"]` in config.cfg?

I guess this sort of works. Is this the recommended way?

```
nlp = spacy.load('en_core_web_md')
senter_nlp.add_pipe("parser", source=nlp)
```

Also what happens to tok2vec during training? Doesn't it get shared across different components?

**Comment by polm** [comment, marked as answer]
Note this issue has a lot of overlap with https://github.com/explosion/spaCy/discussions/7975. 

> Also what happens to tok2vec during training? Doesn't it get shared across different components?

tok2vec is shared by multiple components, but components can also have their own tok2vec. See [the docs](https://spacy.io/usage/training#:~:text=shared%20tok2vec%20listener%20layer) for more info. (This is the same link as in your other thread, but it's here for anyone else who sees this.)

-----

Your steps 1, 2, 3 are fine. Since it sounds like you just want to use an existing pipeline with your custom senter, I would load the existing pipeline and then add your senter the way you're adding the parser in your example code. 

I would also give the senter its own tok2vec as explained at the link.
